NH-3954 - Dynamic proxy cache may yield a wrong proxy

fredericDelaporte opened the pull request:
Fixes [NH-3954](https://nhibernate.jira.com/browse/NH-3954) - Dynamic proxy cache may yield a wrong proxy.

It relies on a dictionary cache, using a key which was implementing equals as hashcode equality.  
But hascode are not unique, and two different objects, even of the same class with different properties/members values, may have the same hashcode. This is documented on [msdn](https://msdn.microsoft.com/en-us/library/zdee4b3y(v=vs.110).aspx).

Quite unlikely, but it happened to me this night. While testing some other changes, `StatelessSessionFetchingTest.DynamicFetch` was unexpectedly failing. And this was because some other proxy got cached with the same hashcode as the proxy required by this test.

hazzik commented on src/NHibernate/Proxy/DynamicProxy/ProxyCacheEntry.cs:
```diff
@@ -43,15 +45,35 @@ public ProxyCacheEntry(System.Type baseType, System.Type[] interfaces)
+			if (_uniqueInterfaces == null)
+				_uniqueInterfaces = new HashSet<System.Type>(Interfaces);
+			if (that._uniqueInterfaces == null)
+				that._uniqueInterfaces = new HashSet<System.Type>(Interfaces);
```
this shall be 

```
that._uniqueInterfaces = new HashSet<System.Type>(that.Interfaces);
```
(note `that.` in front of `Interfaces`)

fredericDelaporte replied:
Damn, I should not code that late...

fredericDelaporte replied:
Fixed, squashed, ... I think this `Serializable` attribute is indeed unneeded for current usage of that class. Removing it would allow a simpler implementation. Maybe I have been too conservative here.

fredericDelaporte commented:
I forget to mention in my PR message: on [Jira side in a comment](https://nhibernate.jira.com/browse/NH-3954?focusedCommentId=49200&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-49200), I have listed a bunch of other classes having the same defect in NHibernate. This PR does not fix them.

fredericDelaporte commented:
The TeamCity failure occured for MySql build, new test failure in `ExpressionSessionLeakTest.SessionGetsCollected`.

It appears that it has not fail on the fisrt build for this PR, executed on code before the fix I have done in response to Hazzik's review.

Due to the last change being very small, I doubt this failure has something to do with the change. Is it a known test case for unexpected failures? 
(Furthermore, debugging the test and setting breakpoint in `ProxyCache`, it appears this test  does not use this feature.)

fredericDelaporte commented:
I have made changes to tests: added direct tests on `Equals`, set cache tests as only run explicitly since a bit unpredictable, and simplified test implementation. (Tests locally re-tested both with original code and fixed code.)

fredericDelaporte commented:
Updated: was re-implementing set equality, used .Net built-in method instead. And re-based.

fredericDelaporte commented:
Rebased for resolving a conflict on test project.

hazzik commented on src/NHibernate/Proxy/DynamicProxy/ProxyCacheEntry.cs:
```diff
@@ -43,15 +44,26 @@ public ProxyCacheEntry(System.Type baseType, System.Type[] interfaces)
+			if (that._uniqueInterfaces == null)
+				that._uniqueInterfaces = new HashSet<System.Type>(that.Interfaces);
+
+			return _uniqueInterfaces.SetEquals(that._uniqueInterfaces);
```
`SetEquals` does not require the argument to be `ISet<>`, so the second if (`if (that._uniqueInterfaces == null)`) is not required

hazzik commented:
There is a possibility that two instances of ProxyCacheEntry will be semantically equal (have the same base type and interfaces) but falsely declared unequal due the fact that set is not ordered.

hazzik commented on src/NHibernate/Proxy/DynamicProxy/ProxyCacheEntry.cs:
```diff
@@ -43,15 +44,26 @@ public ProxyCacheEntry(System.Type baseType, System.Type[] interfaces)
 		public System.Type BaseType { get; private set; }
 		public System.Type[] Interfaces { get; private set; }
```
As we are allowed to break everything in NH5, Please make this `IReadOnlyCollection<System.Type>` without getter.

hazzik replied:
Fixed. I was talking about setter, of cource

hazzik commented on src/NHibernate/Proxy/DynamicProxy/ProxyCacheEntry.cs:
```diff
@@ -43,15 +44,26 @@ public ProxyCacheEntry(System.Type baseType, System.Type[] interfaces)
 		public System.Type BaseType { get; private set; }
```
Please remove getter as per below.

hazzik commented on src/NHibernate/Proxy/DynamicProxy/ProxyCacheEntry.cs:
```diff
@@ -24,6 +24,7 @@ public ProxyCacheEntry(System.Type baseType, System.Type[] interfaces)
 			}
 			BaseType = baseType;
 			Interfaces = interfaces ?? new System.Type[0];
+			_uniqueInterfaces = new HashSet<System.Type>(Interfaces);
```
I think, what we need is presort `Interfaces` (eg. `interfaces.Distinct().OrderBy(x=>x.FullName).ToArray()`).

fredericDelaporte replied:
[`SetEquals`](https://msdn.microsoft.com/en-us/library/dd412096(v=vs.110).aspx) is not sensible to order.

> This method ignores the order of elements and any duplicate elements in other.

So I do not expect any order mismatch to cause an issue.

hazzik replied:
Order mismatch can (?) happen in the calculation of `hashCode`

hazzik replied:
Ok,  no it can not, as hashCode is calculated using just xor (which is strange as there will be lots of collisions)

fredericDelaporte replied:
But there is the hashcode computation too. I am wrong thinking xor is permutable? (Will yield same result whatever the evaluation order.) I am no mure sure about that. But I can not find an exemple for which it would be false.

fredericDelaporte replied:
There is a second HashSet used on the one I have introduced, which is now useless and could be replaced by a simple List. I will change that too.

fredericDelaporte replied:
> Ok, no it can not, as hashCode is calculated using just xor (which is strange as there will be lots of collisions)

When used to combine hashcode from a set (so different objects), it should not be that bad. It is not like `&` or `|` which ends up zeroing or "one-ing" all bits. It seems it is even a [common practice in Java](http://stackoverflow.com/a/19139500/1178314).

Now if we want to change it for a more classical approach like multiplying by a prime then adding next value while overflowing, we will need to take care of the order.

hazzik commented on src/NHibernate/Proxy/DynamicProxy/ProxyCacheEntry.cs:
```diff
@@ -43,15 +44,26 @@ public ProxyCacheEntry(System.Type baseType, System.Type[] interfaces)
 			var that = obj as ProxyCacheEntry;
-			if (ReferenceEquals(null, that))
+			if (ReferenceEquals(null, that) ||
+				hashCode != that.hashCode ||
```
I think `hashCode` check is redundant here, but it's ok to leave it as is.

fredericDelaporte replied:
I left it as an easy early "exit case". It will most of time speed up non equal cases, while being a very minor overhead for equal cases. I should add a comment for warning the hashcode is only good for early false yielding, not for telling it is equal.

fredericDelaporte commented:
> Could you please check the deserialization case?

About this serialization, I was wondering:

> I think this Serializable attribute is indeed unneeded for current usage of that class. Removing it would allow a simpler implementation. Maybe I have been too conservative here.

Is there a known reason for having this class serializable? It may be something inherited from the LinFu library but maybe useless for NHibernate.

fredericDelaporte commented:
A more convoluted test case is added, with many interfaces, different order, duplicates, resolving to same set.

Code changed as suggested (with a bit more c#6/7 syntax, and more comments).

Additionally, I have tested the 5 thousands tests without this `[Serializable]` on `ProxyEntryCache`, no failure. (I have not committed this removal.)

fredericDelaporte commented:
Hum, team city build failed. Maybe using C#7 features is too early. I removed the ["throw expression"](https://stackoverflow.com/documentation/c%23/1936/c-sharp-7-0-features/20215/throw-expressions#t=201703200103502320517).

hazzik commented:
Yep, no C#7 until #569

fredericDelaporte commented:
Rebased (& commit reworded).

fredericDelaporte commented:
Rebased and done.

hazzik commented on src/NHibernate/Proxy/DynamicProxy/ProxyCacheEntry.cs:
```diff
@@ -11,52 +11,57 @@
+			_uniqueInterfaces = new HashSet<System.Type>(interfaces ?? new System.Type[0]);
 
-			if (Interfaces.Length == 0)
+			if (Interfaces.Count == 0)
```
Please change this to `_uniqueInterfaces.Count == 0`

fredericDelaporte replied:
Done.

fredericDelaporte commented:
Well, I just discover following test: `Legacy.MasterDetailTests.Serialize`. It does serialize a whole `ISession`, deserialize it and goes on using it... So if we want to maintain this possibility, cache entries should remain serializable. But I would still keep a simplified implementation of it, the hashtable can be serialized. I would just add back that [Serializable] attribute.

hazzik commented:
Does it fail?

fredericDelaporte commented:
No, but very likely because it does not try to serialize a session having cached some dynamic proxy.

fredericDelaporte commented:
I will try to cause this serialization test to have a cached proxy.

hazzik commented:
The ProxyFactory and its dependencies do not get serialized. It's all fine here.

fredericDelaporte commented:
Ok, great, thanks.
